### PR TITLE
Stress tests

### LINKED ISSUE #16: Stress test adding data to the tree
5ff85ad6ec6fd5214a9edc59232ca4e53a5d6199 is an indication that we should have a test that generates a large number of key/val pairs to set and retrieve from the tree.

### PR DESCRIPTION
**Pull request by Liamsi**
will resolve #16

### CODE REVIEW

**Comment by c633** [comment]
Should we also resolve #18 ?

**Comment by masomel** [comment]
> Should we also resolve #18 ?

I think we should do this in a separate PR since I expect that we'll have to move code around quite a bit for this.

**Comment by Liamsi** [comment]
> Should we also resolve #18 ?

Some of the code will be re-usable to tackle #18. Will try to keep that in mind.

File: crypto/crypto.go [hunk 1]
```diff
@@ -14,7 +14,7 @@ const (
 )
 
 type SigningKey ed25519.PrivateKey
-type VerifKey ed25519.PublicKey
+type VerifyKey ed25519.PublicKey
```
Comment (arlolra) [inline crypto/crypto.go]: This stuff doesn't look like belongs in this pull.

You should `git rebase -i master` and `d` (drop) that commit.

Comment (Liamsi) [inline crypto/crypto.go]: Should be OK now.

File: merkletree/pad_test.go [hunk 2]
```diff
@@ -235,3 +239,194 @@ func TestPoliciesChange(t *testing.T) {
...
+	pad, err := NewPAD(NewPolicies(3, vrfPrivKey1), signKey, 3)
+	if err != nil {
+		t.Fatal(err)
+	}
+	tb, err := pad.TB(key1, val1)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// TODO shouldn't there be a serialize function?
```
Comment (c633) [inline merkletree/pad_test.go]: Can we use this function? https://github.com/coniks-sys/coniks-go/blob/master/merkletree/tb.go#L21-L27

File: merkletree/str.go [hunk 3]
```diff
@@ -27,9 +21,6 @@ type SignedTreeRoot struct {
 }
 
 func NewSTR(key sign.PrivateKey, policies Policies, m *MerkleTree, epoch uint64, prevHash []byte) *SignedTreeRoot {
-	if epoch < 0 {
```
Comment (c633) [inline merkletree/str.go]: Good catch. Thanks!

File: utils/util_test.go [hunk 4]
```diff
@@ -27,3 +28,35 @@ func TestBitsBytesConvert(t *testing.T) {
...
 }
+
+func TestIntToBytes(t *testing.T) {
+	numInt := 42
+	b := IntToBytes(numInt)
+	if int(binary.LittleEndian.Uint32(b)) != numInt {
+		t.Fatal("Conversion to bytes looks wrong!")
+	}
+	// TODO It is possible to call `IntToBytes` with a signed int < 0
+	// isn't that problematic?
```
Comment (masomel) [inline utils/util_test.go]: Good point. I don't expect the server will use signed ints. Do you think ``IntToBytes()`` should throw an error when it receives a negative value?

Comment (Liamsi) [inline utils/util_test.go]: I think to let it `panic` or only use `UIntToBytes` would make sense. As far as I can see `node.level` can only grow (or never is a value subtracted from level). Maybe we could even refactor `level` to be unsigned?

File: merkletree/pad_test.go (hunk 2 above)
Comment (Liamsi) [inline merkletree/pad_test.go]: Makes sense! Thanks for the pointer.

File: utils/util_test.go (hunk 4 above)
Comment (masomel) [inline utils/util_test.go]: Refactoring ``level`` to be unsigned makes a lot of sense to me. The only scenario in which level could decrease is when the tree is re-randomized after the VRF key is changed. But this should still work with an unsigned ``level``.

File: merkletree/pad_test.go (hunk 2 above)
Comment (Liamsi) [inline merkletree/pad_test.go]: Now there is even a public `Serialize` method (https://github.com/coniks-sys/coniks-go/blob/master/merkletree/tb.go#L29-L31). I would prefer to use it instead.

File: utils/util_test.go (hunk 4 above)
Comment (Liamsi) [inline utils/util_test.go]: BTW unsigned inters can overflow in golang without panic/error or something similar: https://golang.org/ref/spec#Integer_overflow

Comment (Liamsi) [inline utils/util_test.go]: > Refactoring level to be unsigned makes a lot of sense to me. The only scenario in which level could decrease is when the tree is re-randomized after the VRF key is changed. But this should still work with an unsigned level.

OK, should this go into another issue/PR?

Comment (c633) [inline utils/util_test.go]: > OK, should this go into another issue/PR?

I think yes.

File: merkletree/pad_test.go (hunk 2 above)
Comment (c633) [inline merkletree/pad_test.go]: Makes sense to me!

File: utils/util_test.go (hunk 4 above)
Comment (Liamsi) [inline utils/util_test.go]: OK, thanks! Opened https://github.com/coniks-sys/coniks-go/issues/53 for that.

**Comment by arlolra** [comment]
Was this done?

**Comment by c633** [comment]
Oops. I definitely did something terrible. Sorry for the noise.

**Comment by c633** [comment]
Please see #54 . I'm so sorry about that :(

**Comment by Liamsi** [comment]
> Oops. I definitely did something terrible. Sorry for the noise.
> Please see #54 . I'm so sorry about that :(

Please, don't worry. Luckily, it easy to revert using git(hub) as you already did :-)

**Pull request merged**